flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/OutputColsHelper.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.common.utils;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.types.Row;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+
+/**
+ * Util for generating output schema when doing prediction or transformation.
+ *
+ * <p>Input:
+ * 1) Schema of input data being predicted or transformed.
+ * 2) Output column names of the prediction/transformation operator.
+ * 3) Output column types of the prediction/transformation operator.
+ * 4) Reserved column names, which is a subset of input data's column names that we want to preserve.
+ *
+ * <p>Output:
+ * 1）The result data schema. The result data is a combination of the preserved columns and the operator's
+ * output columns.
+ *
+ * <p>Several rules are followed:
+ * 1) If reserved columns are not given, then all columns of input data is reserved.
+ * 2）The reserved columns are arranged ahead of the operator's output columns in the final output.
+ * 3) If some of the reserved column names overlap with those of operator's output columns, then the operator's
```

Comment:
I think we should throw an exception when output column names and the user-defined reserved column names have conflict (ultimately both are defined by user, this shouldn't occur on the users' perspective)

Reply:
I think allowing such conflicts is more handy for users, since by default we are reserving every columns.

Reply:
replying to both comments above: the logic IMO assumes too much user behavior: the default reserving all columns in the input does seems convenient. but do you think adding a flag from the input argument (e.g. `reseverAllColumns`) makes more sense?

As a framework I think the responsibility is to strictly track the final result column name uniqueness. By silently drop an input column that users who write a transformation might not have the full knowledge of seems to be an unnecessary responsibility to hand it over to the user.

I would propose:
1. by default reserves no input columns (this is also the standard behavior of a transform in dataset/datastream - makes it consistent with user expectation.
2. create a `inputAsPassthrough` boolean flag when initializing `OutputColsHelper` so that users explicitly have to say "I want all input columns reserved"
3. when `inputAsPassthrough` flag enables; we will take `reservedColumn...` arguments into consideration. 

nit: I prefer the word `passthroughColumn...` over `reserve` makes more sense to me. It would've been nice if some native English speaker can provide more insights on this.

Reply:
Thanks for your advice. As far as we know from our practice, in the domain of ML, reserving input columns by default could be a better design. This is more than the framework itself, but also about user experience. For example, data preprocessing, always the first step to ML, often performs transformation on multiple columns one by one. So there may be PilelineStage_1 for feature scaling on column_1, PilelineStage_2 for binarizing on column_2 and more. This would be redundant to force user to add a flag for each of these steps.

Reply:
I see your point. I still reserves my opinion. but I think one point you made was very crucial: since this is an internal tooling (e.g. end-user does not directly interact with this class) I am ok with this behavior.